src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -269,77 +231,84 @@ internal ConcurrentDictionary(int concurrencyLevel, int capacity, bool growLockA
                 capacity = concurrencyLevel;
             }
 
-            object[] locks = new object[concurrencyLevel];
-            for (int i = 0; i < locks.Length; i++)
+            var locks = new object[concurrencyLevel];
+            locks[0] = locks; // reuse array as the first lock object just to avoid an additional allocation
```

Comment:
If `_locks` is resized, `_locks[0]` will keep the original array alive — unless you assign `newLocks[0] = newLocks` in `GrowTable`.  (Ideally lock objects should be placed on separate cache lines to avoid false sharing.)

Reply:
> If _locks is resized, _locks[0] will keep the original array alive

Yes.  But that only affects the array itself, as it's already reusing the individual lock objects anyway.  So in the 99.9% scenario, it's only keeping alive an extra Environment.ProcessorCount*IntPtr.Size bytes, and if we're resizing, it's because we've grown the dictionary to the point where that should be a drop in the bucket.  On top of that, I almost never see folks explicitly configure the concurrency level, and in cases where they do, it's almost always to 1 or a similarly small value, where they're trying to minimize how many objects are created.  For long-lived and/or large dictionaries, these costs don't matter either way.  The concern is for short-lived / smaller dictionaries, where all these extra objects do start to show up in profiles (we actually decreased the default concurrency level a few releases ago, because of multiple real-world issues with the number of these objects).  This seems like the right trade-off to me.

> Ideally lock objects should be placed on separate cache lines to avoid false sharing.

That should only matter if there's never contention on the objects; once there is contention and the syncblk is expanded, the placement of the object itself in memory shouldn't matter from a false-sharing perspective, since it's the syncblks that'll be read/written, and their placement is managed by the runtime.  On top of that, I'd be loath to increase the size of these objects to 64 bytes or 128 bytes each in order to fill a cache line, for the same reasons discussed above.  If you're interested in experimenting with whether there is false sharing happening in real scenarios, one possible fix would be to lazily-create the lock objects when they're first needed; that _might_ increase the chances of them being created from different threads, in which case they're much more likely to be allocated on (and remain on even with compaction) different cache lines.

Reply:
Thank you for the detailed explanation!